File: backend/components/metiers/administration_publique/tests/utils/rbac/rbac_utils.test.py
# rbac_utils.test.py – Tests unitaires ultra avancés pour rbac_utils.py
import pytest
from .rbac_utils import has_permission, get_user_permissions, PERMISSIONS

def test_admin_permissions():
    user = {'roles': ['admin']}
    for p in PERMISSIONS['admin']:
        assert has_permission(user, p)

def test_auditor_permissions():
    user = {'roles': ['auditor']}
    assert has_permission(user, 'read')
    assert has_permission(user, 'audit')
    assert not has_permission(user, 'write')

def test_multi_roles_permissions():
    user = {'roles': ['editor', 'auditor']}
    assert has_permission(user, 'write')
    assert has_permission(user, 'audit')

def test_get_user_permissions():
    user = {'roles': ['editor', 'auditor']}
    perms = get_user_permissions(user)
    assert 'write' in perms
    assert 'audit' in perms
    assert 'read' in perms

File: backend/components/metiers/administration_publique/tests/utils/rbac/rbac_utils.test.js
// rbac_utils.test.js – Tests unitaires ultra avancés pour rbac_utils.js
const rbac = require('./rbac_utils');

describe('rbac_utils – Permissions', () => {
  it('admin a toutes les permissions', () => {
    const user = { roles: ['admin'] };
    rbac.PERMISSIONS.admin.forEach(p => {
      expect(rbac.hasPermission(user, p)).toBe(true);
    });
  });
  it('auditor peut lire et auditer, mais pas écrire', () => {
    const user = { roles: ['auditor'] };
    expect(rbac.hasPermission(user, 'read')).toBe(true);
    expect(rbac.hasPermission(user, 'audit')).toBe(true);
    expect(rbac.hasPermission(user, 'write')).toBe(false);
  });
  it('multi-rôles hérite de toutes les permissions', () => {
    const user = { roles: ['editor', 'auditor'] };
    expect(rbac.hasPermission(user, 'write')).toBe(true);
    expect(rbac.hasPermission(user, 'audit')).toBe(true);
  });
  it('getUserPermissions retourne toutes les permissions uniques', () => {
    const user = { roles: ['editor', 'auditor'] };
    const perms = rbac.getUserPermissions(user);
    expect(perms).toContain('write');
    expect(perms).toContain('audit');
    expect(perms).toContain('read');
  });
});
